### PR TITLE
Generate function

### PR DESCRIPTION
**Pull request by naterush**
# Description

Adds the Code Options taskpane that allows users to toggle generating functions, and then name their functions. Soon, it will allow them to parameterize these functions!

# Testing

Use it!

# Documentation

No.

### CODE REVIEW

File: mitosheet/mitosheet/step_performers/dataframe_steps/dataframe_rename.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def execute(cls, prev_state: State, params: Dict[str, Any]) -> Tuple[State, Opti
         # Create a new step and save the parameters
         post_state = prev_state.copy()
 
-        post_state.df_names[sheet_index] = get_valid_dataframe_name(post_state.df_names, new_dataframe_name)
+        post_state.df_names[sheet_index] = get_valid_python_identifier(post_state.df_names, new_dataframe_name)
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/dataframe_steps/dataframe_rename.py:51]: Change these back

File: mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+/* 
+    This is the CodeOptions taskpane, allows you to configure how the code is generated
+*/
+const CodeOptionsTaskpane = (props: CodeOptionsTaskpaneProps): JSX.Element => {
+
+    const [codeOptions, setCodeOptions] = useState(() => props.analysisData.codeOptions);
+
+    useDebouncedEffect(() => {
+        props.mitoAPI.updateCodeOptions(codeOptions);
```
Comment (naterush) [inline mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx:33]: Void

File: mitosheet/src/components/Mito.tsx [hunk 3]
```diff
@@ -196,7 +197,8 @@ export const Mito = (props: MitoProps): JSX.Element => {
                 }
 
                 // Then, we replay the analysis to replay!
-                const error = await mitoAPI.updateReplayAnalysis(analysisToReplayName);
+                const error = await mitoAPI.updateReplayAnalysis(analysisToReplayName, args);
+                console.log(error)
```
Comment (naterush) [inline mitosheet/src/components/Mito.tsx:201]: Remove

File: mitosheet/mitosheet/updates/replay_analysis.py [hunk 4]
```diff
@@ -98,17 +102,27 @@ def execute_replay_analysis_update(
         analysis = overwrite_import_data(analysis, step_import_data_list_to_overwrite)
 
     previous_public_interface_version = steps_manager.public_interface_version
+    previous_code_options = steps_manager.code_options
+
+    print("REPLAYING ANALYSIS")
```
Comment (naterush) [inline mitosheet/mitosheet/updates/replay_analysis.py:107]: Remove

File: mitosheet/mitosheet/updates/replay_analysis.py [hunk 5]
```diff
@@ -98,17 +102,27 @@ def execute_replay_analysis_update(
...
     try:
         # Before we execute the steps, update to the public interface version of the saved analysis
         steps_manager.public_interface_version = analysis['public_interface_version']
+        steps_manager.code_options = analysis['code_options']
+        print('replaying analysis with public interface version', analysis['code_options'])
```
Comment (naterush) [inline mitosheet/mitosheet/updates/replay_analysis.py:112]: Remove

**Review by naterush — COMMENTED** [review]
Cleanup!

**Comment by aarondr77** [comment]
Formatting this code better would be nice. 
1. Adding the new lines back
2. Adding a new line before the return 

Screenshot 2023-04-20 at 9 41 34 AM

**Comment by aarondr77** [comment]
1. Create a simple analysis
2. Toggle on generate function
3. Undo all a ton of times
4. Get this generated code. 

Screenshot 2023-04-20 at 9 42 09 AM

**Comment by aarondr77** [comment]
The indenting for the pivot table does not work properly. Notice the pivot table params are not indented and the `)` 

Screenshot 2023-04-20 at 9 48 56 AM

**Comment by aarondr77** [comment]
I suppose this code is not any worse than what we currently have, but it's a wee bit sus with setting overwriting the original variable. 

When I help users generate code like this, I usually firs make a copy of the argument and pass that copy to the function just to keep things simple. Often the user will want to take different slices of that dataframe. ie: get one dataframe that is all of the June data, one that is all of the July data. We end up paramertiziing the function further to take a `June` or `July` that is then a filter condition. 

So tldr: would be down to update from the first screenshot to the second:

Screenshot 2023-04-20 at 10 01 03 AM

Screenshot 2023-04-20 at 10 04 02 AM

**Comment by aarondr77** [comment]
We handle args and imported dataframes differently. Here I passed df3 as an arg and used the dataframe import taskpane for df9. Notice that we only pass df3 as an arg to the function. I think we should pass both of them. It doesn't seem like they should be treated any differently to me. 

Screenshot 2023-04-20 at 10 16 01 AM

File: mitosheet/mitosheet/utils.py [hunk 6]
```diff
@@ -43,6 +43,24 @@ def get_first_unused_dataframe_name(existing_df_names: List[str], new_dataframe_
...
+    valid_words = '_'.join(filter(None, [
+        match.group() for match in re.finditer('\w+', original)
+    ]))
+
+    if len(valid_words) == 0:
+        return default
+    
+    # A valid variable name cannot be empty, or start with a number
+    if valid_words[0].isdecimal():
+        valid_words = prefix + valid_words
```
Comment (aarondr77) [inline mitosheet/mitosheet/utils.py:60]: Nice idea!

File: mitosheet/src/components/Mito.tsx [hunk 7]
```diff
@@ -166,7 +167,7 @@ export const Mito = (props: MitoProps): JSX.Element => {
             // Then, we go and read the arguments to the mitosheet.sheet() call. If there
             // is an analysis to replay, we use this to help lookup the call
             const args = await getArgs(analysisData.analysisToReplay?.analysisName);
-            await mitoAPI.updateArgs(args);
+            console.log("ARGS: ", args)
```
Comment (aarondr77) [inline mitosheet/src/components/Mito.tsx:170]: remove

File: mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            <DefaultTaskpaneBody
+                requiresEnterprise={{
+                    featureName: "code_options",
+                    mitoAPI: props.mitoAPI
+                }}
+                userProfile={props.userProfile}
+            >
+                <Row justify='space-between' align='center'>
+                    <Col>
+                        <LabelAndTooltip tooltip="A Python function is a reusable block of code that performs a specific task. It takes input, processes it, and returns output, making your code more organized and able to be easily rerun on new datasets.">
```
Comment (aarondr77) [inline mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx:51]: When we do the release, let's make a how-to for this feature. We could link the the funciton explanation blog too

File: mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                                // Update directly to avoid the timeout here
+                                void props.mitoAPI.updateCodeOptions(codeOptions);
+                                setCodeOptions(newCodeOptions);
+                            }}
+                        />
+                    </Col>
+                </Row>
+                <Row justify='space-between' align='center'>
+                    <Col>
+                        <LabelAndTooltip tooltip="If you are generating a function, you can give it a short, descriptive name to say what it does.">
```
Comment (aarondr77) [inline mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx:70]: Maybe just "Give your function a short, descriptive name to say what it does"

File: mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                            Function Name
+                        </LabelAndTooltip>
+                    </Col>
+                    <Col>
+                        <Input
+                            disabled={!codeOptions.as_function}
```
Comment (aarondr77) [inline mitosheet/src/components/taskpanes/CodeOptions/CodeOptionsTaskpane.tsx:76]: The disabled css for inputs is pretty horrible right now. Want to spend 5 minute making it better? Adding a border to it would go a long way.

File: mitosheet/src/components/toolbar/ToolbarCodeDropdown.tsx [hunk 11]
```diff
@@ -39,6 +39,7 @@ const ToolbarCodeDropdown = (props: ToolbarCodeDropdownProps): JSX.Element => {
                     : <></>
                 }
                 {makeToolbarDropdownItem(props.actions[ActionEnum.CODESNIPPETS], props.userProfile)}
+                {makeToolbarDropdownItem(props.actions[ActionEnum.CODEOPTIONS], props.userProfile)}
```
Comment (aarondr77) [inline mitosheet/src/components/toolbar/ToolbarCodeDropdown.tsx:42]: Having CodeSnippets and Code Options and AI transformations here is nonsensical. They pretty much have nothing in common. 

Maybe we can call this `Generated Code Options` or `Configure Code`

And also, we could move it to the top of this dropdown since it's the one that makes the most sense here and AI transofmriaton already has a toolbar button.

File: mitosheet/src/types.tsx [hunk 12]
```diff
@@ -67,6 +67,16 @@ export enum UpdateType {
     RenderCountUpdate = 'render_count_update',
     ChecklistUpdate = 'checklist_update',
     UpdateExistingImportUpdate = 'update_existing_import_update',
+    CodeOptionsUpdate = 'code_options_update',
+}
+
+type ParamName = string;
+type ParamValue = string;
+
+export type CodeOptions = {
```
Comment (aarondr77) [inline mitosheet/src/types.tsx:76]: Add a ts_type test to make sure these are in sync with the backend?

File: mitosheet/mitosheet/tests/test_transpile.py [hunk 13]
```diff
@@ -245,3 +246,78 @@ def test_transpile_multiple_pandas_imports_combined(tmp_path):
...
+        "def function():",
+        f"{TAB}txt = pd.read_csv(r'{tmp_file}')",
+        f"{TAB}return txt",
+        "",
+        "txt = function()"
+    ]
+
+def test_transpile_as_function_df_params():
+    mito = create_mito_wrapper(pd.DataFrame({'A': [1]}))
+    mito.argsUpdate(['df1']) # We have to do this manual
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/tests/test_transpile.py:270]: Or just optional arguments to create_mito_wrapper! Good call.

File: mitosheet/mitosheet/types.py [hunk 14]
```diff
@@ -287,6 +291,11 @@ class AITransformFrontendResult(TypedDict):
         modified_dataframes_recons: Dict[str, ModifiedDataframeReconData]
         prints: str
 
+    class CodeOptions(TypedDict):
+        as_function: bool
+        function_name: str
+        function_params: Dict[ParamName, ParamValue]
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/types.py:297]: Yeah it's a mypy thing, condition types are awful.

File: mitosheet/src/types.tsx (hunk 12 above)
Comment (naterush) [reply mitosheet/src/types.tsx:76]: Hard to do with these types.

**Comment by aarondr77** [comment]
I got this error:
0. Create a df in a code cell
1. Pass a df as an arg
2. Import a new csv file
3. Restart your kernel
4. Recreate the df and rerun the mitosheet.sheet(df)

Screenshot 2023-04-20 at 4 28 20 PM
